### PR TITLE
Make last finalized block calculation lazier

### PR DESCRIPTION
**Pull request by KentShikama**
## Overview
<sup>_What this PR does, and why it's needed_</sup>

We previously attempted to update the last finalized block every block but now we only do
so when we are asked what the last finalized block is. This will make the vdag command a
bit slower but make block processing a lot faster. We also move last finalized block calculation
functions to its own object.

### JIRA ticket:
<sup>_Create it if there isn't one already._</sup>

https://rchain.atlassian.net/browse/RCHAIN-3006

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by KentShikama** [comment]
bors try

File: casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala [hunk 1]
```diff
@@ -478,7 +478,7 @@ object BlockAPI {
         tupleSpaceHash = PrettyPrinter.buildStringNoLimit(tsHash),
         tupleSpaceDump = tsDesc,
         timestamp = timestamp,
-        faultTolerance = normalizedFaultTolerance - initialFault,
+        faultTolerance = normalizedFaultTolerance - initialFault, // TODO: Fix
```
Comment (dzajkowski) [inline casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala]: Fix what? :)

Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala]: It is supposed to be a plus sign but I'll need to type up why.

File: casper/src/main/scala/coop/rchain/casper/LastFinalizedBlockCalculator.scala [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import coop.rchain.blockstorage.{BlockDagRepresentation, BlockDagStorage, BlockStore}
+import coop.rchain.casper.util.ProtoUtil
+import coop.rchain.shared.{Cell, Log}
+import coop.rchain.catscontrib.ListContrib
+
+object LastFinalizedBlockCalculator {
+  // TODO: Extract hardcoded fault tolerance threshold
+  private val faultToleranceThreshold = 0f
+
+  def run[F[_]: Monad: Log: Sync: Concurrent: BlockStore: BlockDagStorage: SafetyOracle](
```
Comment (lukasz-golebiewski) [inline casper/src/main/scala/coop/rchain/casper/LastFinalizedBlockCalculator.scala]: Is this rather complex method tested anywhere (possibly on randomly generated input)?

File: casper/src/main/scala/coop/rchain/casper/LastFinalizedBlockCalculator.scala [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+      stateAfter     <- state.read
+      deploysRemoved = initialHistorySize - stateAfter.deployHistory.size
+      _ <- Log[F].info(
+            s"Removed $deploysRemoved deploys from deploy history as we finalized block ${PrettyPrinter
+              .buildString(finalizedChildHash)}."
+          )
+    } yield ()
+
+  /*
+   * On the first pass, block B is finalized if B's main parent block is finalized
```
Comment (lukasz-golebiewski) [inline casper/src/main/scala/coop/rchain/casper/LastFinalizedBlockCalculator.scala]: This comment looks outdated. The method it applies to does not look like it's checking whether the main parent block is finalized

**Comment by lukasz-golebiewski** [comment]
I cannot see where exactly the calculation is made lazier. I guess the one and only commit should be split into two parts. The first would contain only the code extraction to a new object and the second would contain the clue

File: casper/src/main/scala/coop/rchain/casper/LastFinalizedBlockCalculator.scala (hunk 3 above)
Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/LastFinalizedBlockCalculator.scala]: LastFinalizedBlockCalculator.run is only recursively called if there was a block that was finalized.

File: casper/src/main/scala/coop/rchain/casper/LastFinalizedBlockCalculator.scala (hunk 2 above)
Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/LastFinalizedBlockCalculator.scala]: Basically any call to https://github.com/rchain/rchain/pull/2200/files/2c47336efce2a6afe0cd6b56abf7617a642bec36#diff-de23b4200eaecb35ba261ac3caa4d112R202 tests this

For example,

https://github.com/rchain/rchain/blob/0115d5854bcd5e31402b920a4f820b45ebaf0f1f/casper/src/test/scala/coop/rchain/casper/HashSetCasperTest.scala#L1676

https://github.com/rchain/rchain/blob/0115d5854bcd5e31402b920a4f820b45ebaf0f1f/casper/src/test/scala/coop/rchain/casper/HashSetCasperTest.scala#L1687

https://github.com/rchain/rchain/blob/0115d5854bcd5e31402b920a4f820b45ebaf0f1f/casper/src/test/scala/coop/rchain/casper/HashSetCasperTest.scala#L1698

https://github.com/rchain/rchain/blob/0115d5854bcd5e31402b920a4f820b45ebaf0f1f/casper/src/test/scala/coop/rchain/casper/HashSetCasperTest.scala#L1708

How would a randomly generated block dag testing work here?

**Comment by KentShikama** [comment]
> I cannot see where exactly the calculation is made lazier

Instead of being called with every call to addBlock, it is only called when lastFinalizedBlock is called

File: casper/src/main/scala/coop/rchain/casper/LastFinalizedBlockCalculator.scala (hunk 3 above)
Comment (lukasz-golebiewski) [reply casper/src/main/scala/coop/rchain/casper/LastFinalizedBlockCalculator.scala]: You mean if the "main parent block" was finalized? Or any?

File: casper/src/main/scala/coop/rchain/casper/LastFinalizedBlockCalculator.scala (hunk 2 above)
Comment (lukasz-golebiewski) [reply casper/src/main/scala/coop/rchain/casper/LastFinalizedBlockCalculator.scala]: For starters you could generate random DAGs for which it is trivial to determine which blocks should be finalized and check if the algorithm properly marks them as finalized

File: casper/src/main/scala/coop/rchain/casper/LastFinalizedBlockCalculator.scala (hunk 3 above)
Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/LastFinalizedBlockCalculator.scala]: Main parent block

File: casper/src/main/scala/coop/rchain/casper/LastFinalizedBlockCalculator.scala (hunk 2 above)
Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/LastFinalizedBlockCalculator.scala]: If it is a random DAG, then it isn't trivial to determine which blocks should be finalized - only the genesis block can always be trivially finalized.

Comment (lukasz-golebiewski) [reply casper/src/main/scala/coop/rchain/casper/LastFinalizedBlockCalculator.scala]: Ok

**Comment by lukasz-golebiewski** [comment]
What about splitting the commit, so the clue doesn't have to be searched for through the noise?

**Comment by KentShikama** [comment]
By clue do you mean the TODO: fix stuff?

**Comment by lukasz-golebiewski** [comment]
I mean the essence. The place where the calculation is made lazier

**Comment by KentShikama** [comment]
@lukasz-golebiewski Done; PTAL

**Comment by KentShikama** [comment]
bors try

**Pull request merged**